Feature: Support HashMap<_, _, ahash::RandomState>

I am wondering if we can support HashMap<_, _, hash::Random State> from https://docs.rs/ahash/latest/ahash/. pyo3 supports this type as well. BTW, thanks for your effort, and the library's really helpful!

```
error[E0277]: the trait bound `HashMap<i64, char, ahash::RandomState>: PyStubType` is 
not satisfied
   --> crates/deepbiop-fq/src/python.rs:457:15
    |
457 |     id_table: HashMap<i64, char>,
    |               ^^^^^^^^^^^^^^^^^^ the trait `PyStubType` is not implemented for `
HashMap<i64, char, ahash::RandomState>`
    |
    = help: the trait `PyStubType` is implemented for `HashMap<Key, Value>`

error[E0277]: the trait bound `HashMap<std::string::String, Predict, ahash::RandomStat
e>: PyStubType` is not satisfied
    |
459 | ) -> Result<HashMap<String, Predict>> {
    |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `PyStubType` is not implemented 
for `HashMap<std::string::String, Predict, ahash::RandomState>`, which is required by 
`Result<HashMap<std::string::String, Predict, ahash::RandomState>, anyhow::Error>: PyS
tubType`
    |
    = help: the trait `PyStubType` is implemented for `HashMap<Key, Value>`
    = note: required for `Result<HashMap<std::string::String, Predict, ahash::RandomSt
ate>, anyhow::Error>` to implement `PyStubType`

For more information about this error, try `rustc --explain E0277`.

```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -19,6 +19,7 @@ license = "MIT OR Apache-2.0"
 readme = "README.md"
 
 [workspace.dependencies]
+ahash = "0.8.11"
 anyhow = "1.0.86"
 env_logger = "0.11.5"
 insta = "1.39.0"

--- a/examples/pure/Cargo.toml
+++ b/examples/pure/Cargo.toml
@@ -7,6 +7,7 @@ edition.workspace = true
 crate-type = ["cdylib", "rlib"]
 
 [dependencies]
+ahash.workspace = true
 env_logger.workspace = true
 pyo3-stub-gen = { path = "../../pyo3-stub-gen" }
 pyo3.workspace = true

--- a/examples/pure/pure.pyi
+++ b/examples/pure/pure.pyi
@@ -12,6 +12,9 @@ class A:
         ...
 
 
+def ahash_dict() -> dict[str, int]:
+    ...
+
 def create_a(x:int) -> A:
     ...
 

--- a/examples/pure/src/lib.rs
+++ b/examples/pure/src/lib.rs
@@ -1,6 +1,7 @@
 #[cfg_attr(target_os = "macos", doc = include_str!("../../../README.md"))]
 mod readme {}
 
+use ahash::RandomState;
 use pyo3::{exceptions::PyRuntimeError, prelude::*};
 use pyo3_stub_gen::{create_exception, define_stub_info_gatherer, derive::*};
 use std::{collections::HashMap, path::PathBuf};
@@ -74,6 +75,16 @@ fn echo_path(path: PathBuf) -> PyResult<PathBuf> {
     Ok(path)
 }
 
+#[gen_stub_pyfunction]
+#[pyfunction]
+fn ahash_dict() -> HashMap<String, i32, RandomState> {
+    let mut map: HashMap<String, i32, RandomState> = HashMap::with_hasher(RandomState::new());
+    map.insert("apple".to_string(), 3);
+    map.insert("banana".to_string(), 2);
+    map.insert("orange".to_string(), 5);
+    map
+}
+
 /// Initializes the Python module
 #[pymodule]
 fn pure(m: &Bound<PyModule>) -> PyResult<()> {
@@ -85,6 +96,7 @@ fn pure(m: &Bound<PyModule>) -> PyResult<()> {
     m.add_function(wrap_pyfunction!(create_a, m)?)?;
     m.add_function(wrap_pyfunction!(str_len, m)?)?;
     m.add_function(wrap_pyfunction!(echo_path, m)?)?;
+    m.add_function(wrap_pyfunction!(ahash_dict, m)?)?;
     Ok(())
 }
 

--- a/examples/pure/tests/test_python.py
+++ b/examples/pure/tests/test_python.py
@@ -1,4 +1,4 @@
-from pure import sum, create_dict, read_dict, echo_path
+from pure import sum, create_dict, read_dict, echo_path, ahash_dict
 import pytest
 import pathlib
 
@@ -10,6 +10,10 @@ def test_sum():
 
 def test_create_dict():
     assert create_dict(3) == {0: [], 1: [0], 2: [0, 1]}
+
+
+def test_ahash_dict():
+    assert ahash_dict() == {"apple": 3, "banana": 2, "orange": 5}
 
 
 def test_read_dict():

--- a/pyo3-stub-gen/src/stub_type/collections.rs
+++ b/pyo3-stub-gen/src/stub_type/collections.rs
@@ -56,46 +56,49 @@ impl<T: PyStubType> PyStubType for Vec<T> {
     }
 }
 
-macro_rules! impl_map {
-    ($map:ident) => {
-        impl<Key: PyStubType, Value: PyStubType> PyStubType for $map<Key, Value> {
-            fn type_input() -> TypeInfo {
-                let TypeInfo {
-                    name: key_name,
-                    mut import,
-                } = Key::type_input();
-                let TypeInfo {
-                    name: value_name,
-                    import: value_import,
-                } = Value::type_input();
-                import.extend(value_import);
-                import.insert("typing".into());
-                TypeInfo {
-                    name: format!("typing.Mapping[{}, {}]", key_name, value_name),
-                    import,
-                }
+macro_rules! impl_map_inner {
+    () => {
+        fn type_input() -> TypeInfo {
+            let TypeInfo {
+                name: key_name,
+                mut import,
+            } = Key::type_input();
+            let TypeInfo {
+                name: value_name,
+                import: value_import,
+            } = Value::type_input();
+            import.extend(value_import);
+            import.insert("typing".into());
+            TypeInfo {
+                name: format!("typing.Mapping[{}, {}]", key_name, value_name),
+                import,
             }
-            fn type_output() -> TypeInfo {
-                let TypeInfo {
-                    name: key_name,
-                    mut import,
-                } = Key::type_output();
-                let TypeInfo {
-                    name: value_name,
-                    import: value_import,
-                } = Value::type_output();
-                import.extend(value_import);
-                TypeInfo {
-                    name: format!("dict[{}, {}]", key_name, value_name),
-                    import,
-                }
+        }
+        fn type_output() -> TypeInfo {
+            let TypeInfo {
+                name: key_name,
+                mut import,
+            } = Key::type_output();
+            let TypeInfo {
+                name: value_name,
+                import: value_import,
+            } = Value::type_output();
+            import.extend(value_import);
+            TypeInfo {
+                name: format!("dict[{}, {}]", key_name, value_name),
+                import,
             }
         }
     };
 }
 
-impl_map!(HashMap);
-impl_map!(BTreeMap);
+impl<Key: PyStubType, Value: PyStubType> PyStubType for BTreeMap<Key, Value> {
+    impl_map_inner!();
+}
+
+impl<Key: PyStubType, Value: PyStubType, State> PyStubType for HashMap<Key, Value, State> {
+    impl_map_inner!();
+}
 
 macro_rules! impl_tuple {
     ($($T:ident),*) => {

--- a/pyo3-stub-gen/src/stub_type/collections.rs
+++ b/pyo3-stub-gen/src/stub_type/collections.rs
@@ -1,5 +1,5 @@
 use crate::stub_type::*;
-use std::collections::{BTreeMap, HashMap};
+use std::collections::{BTreeMap, BTreeSet, HashMap};
 
 impl<T: PyStubType> PyStubType for Option<T> {
     fn type_input() -> TypeInfo {
@@ -51,6 +51,26 @@ impl<T: PyStubType> PyStubType for Vec<T> {
         let TypeInfo { name, import } = T::type_output();
         TypeInfo {
             name: format!("list[{}]", name),
+            import,
+        }
+    }
+}
+
+impl<T: PyStubType, State> PyStubType for HashSet<T, State> {
+    fn type_output() -> TypeInfo {
+        let TypeInfo { name, import } = T::type_output();
+        TypeInfo {
+            name: format!("set[{}]", name),
+            import,
+        }
+    }
+}
+
+impl<T: PyStubType> PyStubType for BTreeSet<T> {
+    fn type_output() -> TypeInfo {
+        let TypeInfo { name, import } = T::type_output();
+        TypeInfo {
+            name: format!("set[{}]", name),
             import,
         }
     }

--- a/pyo3-stub-gen/src/stub_type/pyo3.rs
+++ b/pyo3-stub-gen/src/stub_type/pyo3.rs
@@ -71,6 +71,7 @@ impl_builtin!(PyList, "list");
 impl_builtin!(PyTuple, "tuple");
 impl_builtin!(PySlice, "slice");
 impl_builtin!(PyDict, "dict");
+impl_builtin!(PySet, "set");
 impl_builtin!(PyString, "str");
 impl_builtin!(PyBackedStr, "str");
 impl_builtin!(PyByteArray, "bytearray");

